dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -0,0 +1,339 @@
+# Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import nvidia.dali.fn as _fn
+import nvidia.dali.pipeline as _pipeline
+import nvidia.dali.tensors as _Tensors
+import nvidia.dali.types as _types
+from nvidia.dali.data_node import DataNode as _DataNode, _arithm_op
+import inspect
+
+
+class DataNodeDebug(_DataNode):
+    """Wrapper class around Tensor, implementing all of the DataNode attributes."""
+
+    def __init__(self, data, name, device, source):
+        super().__init__(name, device, source)
+        self._data = data
+
+    def gpu(self):
+        if self.device == 'gpu':
+            return self
+        return DataNodeDebug(self._data._as_gpu(), self.name, 'gpu', self.source)
+
+    def get(self):
+        return self._data
```

Comment:
How about overloading `__getitem__`?
```python
def __getitem__(self, index):
  return self._data[index]
```
This way we might have more seamless debugging:
```python
@nvidia.dali.experimental.pipeline_def(debug=True)
def my_pipe():
    data, _ = fn.readers.file(file_root=images_dir)
    img = fn.decoders.image(data)
    print(np.array(img[0]))  # Instead of img.get()[0]
    ...
```

Reply:
We already use getitem four slicing , so we can't use it here. 
We plan to add better repr and str to DataNode[Debug]

Reply:
Yes, it would change the behavior when converting from regular to debug mode.